[inductor][cpp] bf16/fp16 gemm template computed with fp32 w/o epilogue fusion

jgong5 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #127195
* #127505
* #127687
* #126545
* __->__ #126068

As part of #125683, this PR adds the initial bf16/fp16 gemm template support with micro-gemm implemented with fused type casting and fp32 computation. It doesn't provide epilogue fusion support yet which will be added in the next PR.

cc @voznesenskym @penguinwu @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

Differential Revision: [D58017580](https://our.internmc.facebook.com/intern/diff/D58017580)

Commit:
Update

[ghstack-poisoned]

Commit:
Update

[ghstack-poisoned]

Commit:
Update

[ghstack-poisoned]

Commit:
Update

[ghstack-poisoned]

Commit:
Update on "[inductor][cpp] bf16/fp16 gemm template computed with fp32 w/o epilogue fusion"


This PR adds the initial bf16/fp16 gemm template support with micro-gemm implemented with fused type casting and fp32 computation. It doesn't provide epilogue fusion support yet which will be added in the next PR.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Commit:
Update on "[inductor][cpp] bf16/fp16 gemm template computed with fp32 w/o epilogue fusion"


This PR adds the initial bf16/fp16 gemm template support with micro-gemm implemented with fused type casting and fp32 computation. It doesn't provide epilogue fusion support yet which will be added in the next PR.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Commit:
Update on "[inductor][cpp] bf16/fp16 gemm template computed with fp32 w/o epilogue fusion"


This PR adds the initial bf16/fp16 gemm template support with micro-gemm implemented with fused type casting and fp32 computation. It doesn't provide epilogue fusion support yet which will be added in the next PR.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Commit:
Update on "[inductor][cpp] bf16/fp16 gemm template computed with fp32 w/o epilogue fusion"


This PR adds the initial bf16/fp16 gemm template support with micro-gemm implemented with fused type casting and fp32 computation. It doesn't provide epilogue fusion support yet which will be added in the next PR.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Commit:
Update on "[inductor][cpp] bf16/fp16 gemm template computed with fp32 w/o epilogue fusion"


This PR adds the initial bf16/fp16 gemm template support with micro-gemm implemented with fused type casting and fp32 computation. It doesn't provide epilogue fusion support yet which will be added in the next PR.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jgong5 commented:
@pytorchbot merge

jgong5 commented:
@pytorchbot merge

Commit:
Update on "[inductor][cpp] bf16/fp16 gemm template computed with fp32 w/o epilogue fusion"


As part of #125683, this PR adds the initial bf16/fp16 gemm template support with micro-gemm implemented with fused type casting and fp32 computation. It doesn't provide epilogue fusion support yet which will be added in the next PR.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Commit:
Update on "[inductor][cpp] bf16/fp16 gemm template computed with fp32 w/o epilogue fusion"


As part of #125683, this PR adds the initial bf16/fp16 gemm template support with micro-gemm implemented with fused type casting and fp32 computation. It doesn't provide epilogue fusion support yet which will be added in the next PR.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Commit:
Update on "[inductor][cpp] bf16/fp16 gemm template computed with fp32 w/o epilogue fusion"


As part of #125683, this PR adds the initial bf16/fp16 gemm template support with micro-gemm implemented with fused type casting and fp32 computation. It doesn't provide epilogue fusion support yet which will be added in the next PR.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jgong5 commented:
@pytorchbot merge

Commit:
Update on "[inductor][cpp] bf16/fp16 gemm template computed with fp32 w/o epilogue fusion"


As part of #125683, this PR adds the initial bf16/fp16 gemm template support with micro-gemm implemented with fused type casting and fp32 computation. It doesn't provide epilogue fusion support yet which will be added in the next PR.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Commit:
Update on "[inductor][cpp] bf16/fp16 gemm template computed with fp32 w/o epilogue fusion"


As part of #125683, this PR adds the initial bf16/fp16 gemm template support with micro-gemm implemented with fused type casting and fp32 computation. It doesn't provide epilogue fusion support yet which will be added in the next PR.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jgong5 commented:
@pytorchbot merge

chunyuan-w commented on torch/_inductor/ir.py:
```diff
@@ -6321,17 +6321,17 @@ def codegen(self, wrapper):
 
     @classmethod
-    def create(cls, x, w, b, attr, scalars, algorithm):
+    def create(cls, x, w, B, attr, scalars, algorithm):
```
Hi @jgong5 it seems the `LinearBinary` class also need similar change (from `b` to `B`). I met an issue  [here](https://github.com/pytorch/pytorch/pull/124350#issuecomment-2130665645) and I've added this change to the `LinearBinary` class as https://github.com/pytorch/pytorch/commit/39ed894ea6944b4f24999f146484bcecb4812e1e in my PR to make CI pass there.

jgong5 replied:
Thanks for the fix. Weird that it was not exposed from previous UTs.

Commit:
Update

[ghstack-poisoned]

Commit:
Update

[ghstack-poisoned]

Commit:
Update

[ghstack-poisoned]

jgong5 commented:
@pytorchbot rebase

jgong5 commented:
@pytorchbot merge

PaliC commented:
@pytorchbot revert -m "failing internal tests" -c "ghfirst"

The cpu_select_algorithm tests you added/changed seem to be failing internally

Example of 1 error: https://pastebin.com/BB8E1Ngx

Generally the succinct lines are in the form of 
```
222/torch/include/ATen/cpu/vec/vec_half.h:19:10: error: '__builtin_ia32_vcvtps2ph' needs target feature f16c

616/torch/include/ATen/cpu/vec/vec256/vec256_bfloat16.h:107:7: error: always_inline function '_mm256_cvtph_ps' requires target feature 'f16c', but would be inlined into function 'cvtfp16_fp32' that is compiled without support for 'f16c'

34/torch/include/ATen/cpu/vec/vec_half.h:37:10: error: always_inline function '_cvtsh_ss' requires target feature 'f16c', but would be inlined into function 'half2float_scalar' that is compiled without support for 'f16c'

etc.
```

Commit:
Update on "[inductor][cpp] bf16/fp16 gemm template computed with fp32 w/o epilogue fusion"


As part of #125683, this PR adds the initial bf16/fp16 gemm template support with micro-gemm implemented with fused type casting and fp32 computation. It doesn't provide epilogue fusion support yet which will be added in the next PR.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Commit:
Update on "[inductor][cpp] bf16/fp16 gemm template computed with fp32 w/o epilogue fusion"


As part of #125683, this PR adds the initial bf16/fp16 gemm template support with micro-gemm implemented with fused type casting and fp32 computation. It doesn't provide epilogue fusion support yet which will be added in the next PR.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jgong5 commented:
> @pytorchbot revert -m "failing internal tests" -c "ghfirst"
> 
> The cpu_select_algorithm tests you added/changed seem to be failing internally
> 
> Example of 1 error: https://pastebin.com/BB8E1Ngx
> 
> Generally the succinct lines are in the form of
> 
> ```
> 222/torch/include/ATen/cpu/vec/vec_half.h:19:10: error: '__builtin_ia32_vcvtps2ph' needs target feature f16c
> 
> 616/torch/include/ATen/cpu/vec/vec256/vec256_bfloat16.h:107:7: error: always_inline function '_mm256_cvtph_ps' requires target feature 'f16c', but would be inlined into function 'cvtfp16_fp32' that is compiled without support for 'f16c'
> 
> 34/torch/include/ATen/cpu/vec/vec_half.h:37:10: error: always_inline function '_cvtsh_ss' requires target feature 'f16c', but would be inlined into function 'half2float_scalar' that is compiled without support for 'f16c'
> 
> etc.
> ```

Thanks. I had a fix in another PR on top of it but it seems I should have moved the fix in this PR instead.

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58017580).

Commit:
Update on "[inductor][cpp] bf16/fp16 gemm template computed with fp32 w/o epilogue fusion"


As part of #125683, this PR adds the initial bf16/fp16 gemm template support with micro-gemm implemented with fused type casting and fp32 computation. It doesn't provide epilogue fusion support yet which will be added in the next PR.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

Differential Revision: [D58017580](https://our.internmc.facebook.com/intern/diff/D58017580)

[ghstack-poisoned]

Commit:
Update on "[inductor][cpp] bf16/fp16 gemm template computed with fp32 w/o epilogue fusion"


As part of #125683, this PR adds the initial bf16/fp16 gemm template support with micro-gemm implemented with fused type casting and fp32 computation. It doesn't provide epilogue fusion support yet which will be added in the next PR.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

Differential Revision: [D58017580](https://our.internmc.facebook.com/intern/diff/D58017580)

[ghstack-poisoned]

Commit:
Update on "[inductor][cpp] bf16/fp16 gemm template computed with fp32 w/o epilogue fusion"


As part of #125683, this PR adds the initial bf16/fp16 gemm template support with micro-gemm implemented with fused type casting and fp32 computation. It doesn't provide epilogue fusion support yet which will be added in the next PR.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

Differential Revision: [D58017580](https://our.internmc.facebook.com/intern/diff/D58017580)

[ghstack-poisoned]

Commit:
Update on "[inductor][cpp] bf16/fp16 gemm template computed with fp32 w/o epilogue fusion"


As part of #125683, this PR adds the initial bf16/fp16 gemm template support with micro-gemm implemented with fused type casting and fp32 computation. It doesn't provide epilogue fusion support yet which will be added in the next PR.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

Differential Revision: [D58017580](https://our.internmc.facebook.com/intern/diff/D58017580)

[ghstack-poisoned]

Commit:
Update

[ghstack-poisoned]

desertfire commented:
> @pytorchbot revert -m "failing internal tests" -c "ghfirst"
> 
> The cpu_select_algorithm tests you added/changed seem to be failing internally
> 
> Example of 1 error: [pastebin.com/BB8E1Ngx](https://pastebin.com/BB8E1Ngx)
> 
> Generally the succinct lines are in the form of
> 
> ```
> 222/torch/include/ATen/cpu/vec/vec_half.h:19:10: error: '__builtin_ia32_vcvtps2ph' needs target feature f16c
> 
> 616/torch/include/ATen/cpu/vec/vec256/vec256_bfloat16.h:107:7: error: always_inline function '_mm256_cvtph_ps' requires target feature 'f16c', but would be inlined into function 'cvtfp16_fp32' that is compiled without support for 'f16c'
> 
> 34/torch/include/ATen/cpu/vec/vec_half.h:37:10: error: always_inline function '_cvtsh_ss' requires target feature 'f16c', but would be inlined into function 'half2float_scalar' that is compiled without support for 'f16c'
> 
> etc.
> ```

@PaliC ,  fbcode and github are out of sync on this one. This PR is reverted but D58015598 isn't. Please help to solve the problem.

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58017580).

Commit:
Update

[ghstack-poisoned]

jgong5 commented:
@pytorchbot merge

Commit:
Update

[ghstack-poisoned]

jgong5 commented:
@pytorchbot merge

Commit:
Update

[ghstack-poisoned]

jgong5 commented:
@pytorchbot merge

Commit:
Update

[ghstack-poisoned]

Commit:
Update

[ghstack-poisoned]
